### PR TITLE
Adding support for cross-cluster storage migration for managed storage when using XenServer

### PR DESCRIPTION
**Pull request by mike-tutkowski**
This PR adds support for cross-cluster storage migration of VMs that make use of managed storage with XenServer.

Managed storage is when you have a 1:1 mapping between a virtual disk and a volume on a SAN (in the case of XenServer, an SR is placed on this SAN volume and a single virtual disk placed in the SR).

Managed storage allows features such as storage QoS and SAN-side snapshots to work (sort of analogous to VMware VVols).

This PR focuses on enabling VMs that are using managed storage to be migrated across XenServer clusters.

I have successfully run the following tests on this branch:

TestVolumes.py
TestSnapshots.py
TestVMSnapshots.py
TestAddRemoveHosts.py
TestVMMigrationWithStorage.py (which is a new test that is being added with this PR)

### CODE REVIEW

File: engine/api/src/org/apache/cloudstack/engine/subsystem/api/storage/PrimaryDataStoreDriver.java [hunk 1]
```diff
@@ -25,6 +25,13 @@
 import com.cloud.storage.StoragePool;
 
 public interface PrimaryDataStoreDriver extends DataStoreDriver {
+    String BASIC_CREATE = "basicCreate";
+    String BASIC_DELETE = "basicDelete";
+    String BASIC_DELETE_FAILURE = "basicDeleteFailure";
+    String BASIC_GRANT_ACCESS = "basicGrantAccess";
+    String BASIC_REVOKE_ACCESS = "basicRevokeAccess";
+    String BASIC_IQN = "basicIqn";
```
Comment (syed) [inline engine/api/src/org/apache/cloudstack/engine/subsystem/api/storage/PrimaryDataStoreDriver.java]: @mike-tutkowski What are basic operations? are they different from normal operations?

File: engine/orchestration/src/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 2]
```diff
@@ -2045,62 +2045,74 @@ protected void migrate(final VMInstanceVO vm, final long srcHostId, final Deploy
...
-                    // Volume needs to be migrated. Pick the first pool from the list. Add a mapping to migrate the
-                    // volume to a pool only if it is required; that is the current pool on which the volume resides
-                    // is not available on the destination host.
-                    final Iterator<StoragePool> iter = poolList.iterator();
-                    while (iter.hasNext()) {
-                        if (currentPool.getId() == iter.next().getId()) {
-                            currentPoolAvailable = true;
-                            break;
+                    final DiskProfile diskProfile = new DiskProfile(volume, diskOffering, profile.getHypervisorType());
+                    final DataCenterDeployment plan = new DataCenterDeployment(host.getDataCenterId(), host.getPodId(), host.getClusterId(), host.getId(), null, null);
```
Comment (syed) [inline engine/orchestration/src/com/cloud/vm/VirtualMachineManagerImpl.java]: @mike-tutkowski is the assumption here that both `src` and `dest` volume pools are the same/use the same tags? Would migration between two different primary managed storage using different tags be possible?

File: plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xen610/XenServer610MigrateWithStorageReceiveCommandWrapper.java [hunk 3]
```diff
@@ -56,7 +55,7 @@
     public Answer execute(final MigrateWithStorageReceiveCommand command, final XenServer610Resource xenServer610Resource) {
         final Connection connection = xenServer610Resource.getConnection();
         final VirtualMachineTO vmSpec = command.getVirtualMachine();
-        final List<Pair<VolumeTO, StorageFilerTO>> volumeToFiler = command.getVolumeToFiler();
+        final List<Pair<VolumeTO, String>> volumeToStorageUuid = command.getVolumeToStorageUuid();
```
Comment (syed) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xen610/XenServer610MigrateWithStorageReceiveCommandWrapper.java]: @mike-tutkowski whould changing the map from `StorageFilerTO` to `String` change how other hypervisors like `KVM` would implement this?

File: plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xenbase/CitrixCreateStoragePoolCommandWrapper.java [hunk 4]
```diff
@@ -39,20 +41,35 @@
...
+
         try {
-            if (pool.getType() == StoragePoolType.NetworkFilesystem) {
-                citrixResourceBase.getNfsSR(conn, Long.toString(pool.getId()), pool.getUuid(), pool.getHost(), pool.getPath(), pool.toString());
-            } else if (pool.getType() == StoragePoolType.IscsiLUN) {
-                citrixResourceBase.getIscsiSR(conn, pool.getUuid(), pool.getHost(), pool.getPath(), null, null, false);
-            } else if (pool.getType() == StoragePoolType.PreSetup) {
-            } else {
-                return new Answer(command, false, "The pool type: " + pool.getType().name() + " is not supported.");
+            if (command.getCreateDatastore()) {
```
Comment (syed) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xenbase/CitrixCreateStoragePoolCommandWrapper.java]: can we come here for non-managed storage?

File: plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xenbase/CitrixDeleteStoragePoolCommandWrapper.java [hunk 5]
```diff
@@ -32,22 +34,40 @@
...
             final Answer answer = new Answer(command, true, "success");
+
             return answer;
         } catch (final Exception e) {
-            final String msg = "DeleteStoragePoolCommand XenAPIException:" + e.getMessage() + " host:" + citrixResourceBase.getHost().getUuid() + " pool: " + poolTO.getHost()
-                    + poolTO.getPath();
+            final String msg = "DeleteStoragePoolCommand XenAPIException:" + e.getMessage() + " host:" + citrixResourceBase.getHost().getUuid() +
+                    " pool: " + poolTO.getHost() + poolTO.getPath();
+
             s_logger.warn(msg, e);
```
Comment (syed) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xenbase/CitrixDeleteStoragePoolCommandWrapper.java]: Can we change this `warn` to an `error`?

**Comment by syed** [comment]
@mike-tutkowski I've reviewed the code. it LGTM, my only concern is of migrating between two different managed storage when they don't have same tags (which would ideally be the default)

File: engine/api/src/org/apache/cloudstack/engine/subsystem/api/storage/PrimaryDataStoreDriver.java (hunk 1 above)
Comment (mike-tutkowski) [inline engine/api/src/org/apache/cloudstack/engine/subsystem/api/storage/PrimaryDataStoreDriver.java]: These are like what we did with snapshots and cloning in a previous PR: Ways of passing info into the storage driver to have it perform actions that aren't covered by the current interface.

For example, to make this PR work, we need to have two backend volumes exist for the same CloudStack volume simultaneously (during the copy of the data).

"basicCreate" in volume_details for a given CloudStack volume ID tells the storage driver to simply create a volume on the backend with the same characteristics as the existing volume on the backend (for this CloudStack volume).

We then communicate info back to the storage framework (for XenMotion/vMotion purposes) by adding "basicIqn" to the volume details.

File: engine/orchestration/src/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 2 above)
Comment (mike-tutkowski) [inline engine/orchestration/src/com/cloud/vm/VirtualMachineManagerImpl.java]: For this PR, the limitation is not just storage tagging, but also the src and dest have to be the same zone-wide primary storage.

I know you and I have another task on our shared to-do list to enable cross-primary storage migration (where managed storage is one or both of the primary storages). This, however, does not implement that functionality.

This PR just enables Storage XenMotion/vMotion from one cluster to another where managed storage is involved (and the data effectively stays on the same primary storage).

File: plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xen610/XenServer610MigrateWithStorageReceiveCommandWrapper.java (hunk 3 above)
Comment (mike-tutkowski) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xen610/XenServer610MigrateWithStorageReceiveCommandWrapper.java]: I checked where this was being accessed from throughout the codebase and all hypervisors should be OK with this change.

They all seemed to only care about a string UUID in the StorageFilerTO object.

This PR is only for XenMotion and vMotion. This PR does not enabled cross-cluster storage migration (for managed storage) on any other hypervisor(s) in CloudStack.

File: plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xenbase/CitrixCreateStoragePoolCommandWrapper.java (hunk 4 above)
Comment (mike-tutkowski) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xenbase/CitrixCreateStoragePoolCommandWrapper.java]: Yes, CloudStack does allow an admin to specify an IQN (and other storage info) when adding primary storage that leads to the creation of an SR for you.

To my knowledge, this is not the case on, say, ESXi/vSphere, though. In that case, you have to create your datastore before adding it to CloudStack as primary storage.

File: plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xenbase/CitrixDeleteStoragePoolCommandWrapper.java (hunk 5 above)
Comment (mike-tutkowski) [inline plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/wrapper/xenbase/CitrixDeleteStoragePoolCommandWrapper.java]: Sure, sounds good.

**Comment by mike-tutkowski** [comment]
Also, here's a (roughly) 30-minute video I made that demonstrates this PR in action: https://www.youtube.com/watch?v=_pbnvQxTAqw&list=PLqOXKM0Bt13DFnQnwUx8ZtJzoyDV0Uuye&index=17

**Comment by karuturi** [comment]
@mike-tutkowski I started our internal CI run on this to run the smoke test suite. It will post the results here.

**Comment by mike-tutkowski** [comment]
Thank you, Rajani!

On Sep 12, 2016, at 11:41 PM, Rajani Karuturi <notifications@github.com<mailto:notifications@github.com>> wrote:


@mike-tutkowski<https://github.com/mike-tutkowski> I started our internal CI run on this to run the smoke test suite. It will post the results here.

-

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 92
 Hypervisor xenserver
 NetworkType Advanced
 Passed=102
 Failed=1
 Skipped=4

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* test_non_contigiousvlan.py

 * test_extendPhysicalNetworkVlan Failing since 10 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_ssvm.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_disk_offerings.py

**Comment by karuturi** [comment]
the failed test(test_extendPhysicalNetworkVlan) is not due to this PR. Its a test case cleanup issue. This PR looks good.

**Pull request merged**